src/vcpkg/binarycaching.cpp

```diff
@@ -1419,29 +1422,36 @@ namespace
             return msg::format(msgRestoredPackagesFromAZUPKG, msg::count = count, msg::elapsed = ElapsedTime(elapsed));
         }
 
-        void fetch(View<const InstallPlanAction*> actions, Span<RestoreResult> out_status) const override
+        void acquire_zips(View<const InstallPlanAction*> actions,
+            Span<Optional<ZipResource>> out_zips) const override
         {
             for (size_t i = 0; i < actions.size(); ++i)
             {
-                auto info = BinaryPackageReadInfo{*actions[i]};
-                auto ref = make_feedref(info, "");
-                auto res = m_azure_tool.download(m_source, ref.id, ref.version, info.package_dir, m_sink);
+                const auto& action = *actions[i];
+                const auto info = BinaryPackageReadInfo{action};
+                const auto ref = make_feedref(info, "");
 
-                if (res)
+                Path temp_dir = m_buildtrees / fmt::format("upkg_download_{}", info.package_abi);
+                Path zip_path = temp_dir / fmt::format("{}.zip", ref.id);
+
+                const auto result = m_azure_tool.download(m_source, ref.id, ref.version, temp_dir, m_sink);
+                if (result.has_value() && m_fs.exists(zip_path, IgnoreErrors{}))
                 {
-                    out_status[i] = RestoreResult::restored;
+                    out_zips[i].emplace(std::move(zip_path), RemoveWhen::always);
```

Comment:
```suggestion
                    out_zips[i].emplace(std::move(zip_path), RemoveWhen::nothing);
```

??

Reply:
I think it should be `always`; after we downloaded the zip and extracted it, we don't need the zip anymore. `nothing` is for the files provider because there we extract the zip in place rather than copying it locally first.

Reply:
Actually given the statement above about temp dir the existing RemoveWhen mechanism may be insufficient since there is now an extra temporary directory that has to get removed.